receiver/signalfxreceiver/signalfxv2_to_metricdata.go

```diff
@@ -17,214 +17,190 @@ package signalfxreceiver
 	// 	data point.
-	numDroppedTimeSeries := 0
-	md := consumerdata.MetricsData{}
-	metrics := make([]*metricspb.Metric, 0, len(sfxDataPoints))
+	numDroppedDataPoints := 0
+	md := pdata.NewMetrics()
+	md.ResourceMetrics().Resize(1)
+	rm := md.ResourceMetrics().At(0)
+	rm.InitEmpty()
+
+	rm.InstrumentationLibraryMetrics().Resize(1)
+	ilm := rm.InstrumentationLibraryMetrics().At(0)
+	ilm.InitEmpty()
+
+	metrics := ilm.Metrics()
+	metrics.Resize(len(sfxDataPoints))
+
+	i := 0
 	for _, sfxDataPoint := range sfxDataPoints {
 		if sfxDataPoint == nil {
 			// TODO: Log or metric for this odd ball?
 			continue
 		}
 
+		m := metrics.At(i)
 		// First check if the type is convertible and the data point is consistent.
-		metricType, err := convertType(sfxDataPoint)
+		err := fillInType(sfxDataPoint, m)
 		if err != nil {
-			numDroppedTimeSeries++
+			numDroppedDataPoints++
 			logger.Debug("SignalFx data-point type conversion error",
 				zap.Error(err),
 				zap.String("metric", sfxDataPoint.GetMetric()))
 			continue
 		}
-		point, err := buildPoint(sfxDataPoint, metricType)
+
+		m.SetName(sfxDataPoint.Metric)
+
+		switch m.DataType() {
+		case pdata.MetricDataTypeIntGauge:
+			err = fillIntDataPoint(sfxDataPoint, m.IntGauge().DataPoints())
+		case pdata.MetricDataTypeIntSum:
+			err = fillIntDataPoint(sfxDataPoint, m.IntSum().DataPoints())
+		case pdata.MetricDataTypeDoubleGauge:
+			err = fillDoubleDataPoint(sfxDataPoint, m.DoubleGauge().DataPoints())
+		case pdata.MetricDataTypeDoubleSum:
+			err = fillDoubleDataPoint(sfxDataPoint, m.DoubleSum().DataPoints())
+		}
+
 		if err != nil {
-			numDroppedTimeSeries++
+			numDroppedDataPoints++
 			logger.Debug("SignalFx data-point datum conversion error",
 				zap.Error(err),
 				zap.String("metric", sfxDataPoint.GetMetric()))
 			continue
 		}
 
-		labelKeys, labelValues := buildLabelKeysAndValues(sfxDataPoint.Dimensions)
-		descriptor := buildDescriptor(sfxDataPoint, labelKeys, metricType)
-		ts := &metricspb.TimeSeries{
-			// TODO: StartTimestamp can be set if each cumulative time series are
-			//  	tracked but right now it is not clear if it brings benefits.
-			//		Perhaps as an option so cost is "pay for play".
-			LabelValues: labelValues,
-			Points:      []*metricspb.Point{point},
-		}
-		metric := &metricspb.Metric{
-			MetricDescriptor: descriptor,
-			Timeseries:       []*metricspb.TimeSeries{ts},
-		}
-		metrics = append(metrics, metric)
+		i++
 	}
 
-	md.Metrics = metrics
-	return md, numDroppedTimeSeries
+	metrics.Resize(i)
+
+	return md, numDroppedDataPoints
 }
 
-func convertType(
+func fillInType(
 	sfxDataPoint *sfxpb.DataPoint,
-) (descType metricspb.MetricDescriptor_Type, err error) {
-
+	m pdata.Metric,
+) (err error) {
 	// Combine metric type with the actual data point type
 	sfxMetricType := sfxDataPoint.GetMetricType()
 	sfxDatum := sfxDataPoint.Value
 	if sfxDatum.IntValue == nil && sfxDatum.DoubleValue == nil && sfxDatum.StrValue == nil {
```

Comment:
When can the value be "String"? How do we convert that?

Reply:
Never, got rid of consideration of string and it will return an error if present.